zkevm-circuits/src/evm_circuit/execution/block_ctx.rs

```diff
@@ -0,0 +1,238 @@
+
+        self.value_u64.value.assign(
+            region,
+            offset,
+            Some(u64::try_from(value).unwrap().to_le_bytes()),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct BlockCtxU160Gadget<F> {
+    value_u160: BlockCtxGadget<F, N_BYTES_ACCOUNT_ADDRESS>,
+}
+
+impl<F: Field> ExecutionGadget<F> for BlockCtxU160Gadget<F> {
+    const NAME: &'static str = "BlockCTXU160";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::BLOCKCTXU160;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let value_u160 = BlockCtxGadget::construct(cb);
+
+        Self { value_u160 }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.value_u160
+            .same_context
+            .assign_exec_step(region, offset, step)?;
+
+        let value = block.rws[step.rw_indices[0]].stack_value();
+
+        self.value_u160.value.assign(
+            region,
+            offset,
+            Some(
+                value.to_le_bytes()[..N_BYTES_ACCOUNT_ADDRESS]
+                    .try_into()
+                    .unwrap(),
+            ),
+        )?;
+
+        Ok(())
+    }
+}
+
+#[derive(Clone, Debug)]
+pub(crate) struct BlockCtxU256Gadget<F> {
+    value_u256: BlockCtxGadget<F, N_BYTES_WORD>,
+}
+
+impl<F: Field> ExecutionGadget<F> for BlockCtxU256Gadget<F> {
+    const NAME: &'static str = "BLOCKCTXU256";
+
+    const EXECUTION_STATE: ExecutionState = ExecutionState::BLOCKCTXU256;
+
+    fn configure(cb: &mut ConstraintBuilder<F>) -> Self {
+        let value_u256 = BlockCtxGadget::construct(cb);
+
+        Self { value_u256 }
+    }
+
+    fn assign_exec_step(
+        &self,
+        region: &mut Region<'_, F>,
+        offset: usize,
+        block: &Block<F>,
+        _: &Transaction,
+        _: &Call,
+        step: &ExecStep,
+    ) -> Result<(), Error> {
+        self.value_u256
+            .same_context
+            .assign_exec_step(region, offset, step)?;
+
+        let value = block.rws[step.rw_indices[0]].stack_value();
+
+        self.value_u256
+            .value
+            .assign(region, offset, Some(value.to_le_bytes()))?;
+
+        Ok(())
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use crate::test_util::run_test_circuits;
+    use eth_types::bytecode;
+    use mock::TestContext;
+
+    fn test_ok(bytecode: bytecode::Bytecode) {
+        assert_eq!(
+            run_test_circuits(
+                TestContext::<2, 1>::simple_ctx_with_bytecode(bytecode).unwrap(),
+                None
+            ),
+            Ok(())
+        );
+    }
+
+    #[test]
+    fn blockcxt_u64_gadget_test() {
+        let bytecode = bytecode! {
+            TIMESTAMP
+            STOP
+            NUMBER
+            STOP
+            GASLIMIT
+            STOP
+        };
```

Comment:
You seem to want `POP` instead of `STOP` right?

Reply:
yeah, nice catch!

Reply:
fix block gaslimit cases for trace